### PR TITLE
job: close quickfix window if a job has succeeded

### PR DESCRIPTION
**Pull request by pborzenkov**
Otherwise all previous build errors are shown even after successful
build.

Signed-off-by: Pavel Borzenkov <pavel.borzenkov@gmail.com>

### CODE REVIEW

File: autoload/go/job.vim [hunk 1]
```diff
@@ -57,16 +57,17 @@ function go#job#Spawn(args)
       endif
     endif
 
+    let l:listtype = "quickfix"
```
Comment (fatih) [inline autoload/go/job.vim]: I think we should use here:  `let l:listtype = go#list#Type("quickfix")`. the `Type()` function returns the user preference (can be set via `go_list_type`) , if there is none, then the argument itself is returned.

**Comment by fatih** [comment]
Btw, the current code is all set to `quickfix`, not sure how this fixes it. What exactly is the problem? Is this fixing #1109, if yes what's the real problem here?

**Comment by pborzenkov** [comment]
Yes, I am experiencing the very same problem as described in #1109. And it does indeed fix it for me.

Initially, I though that go#list#Window(0) would close 'localtionlist', but now I see that it should close 'quickfix' by default. So I'm not sure how this patch fixes it, anymore :) I'm going to dig a bit deeper here.

**Comment by pborzenkov** [comment]
Ok, so I added some debug messages to list.vim:

```
--- a/autoload/go/list.vim
+++ b/autoload/go/list.vim
@@ -7,15 +7,20 @@ endif
 " closes the window
 function! go#list#Window(listtype, ...) abort
   let l:listtype = go#list#Type(a:listtype)
+  echom a:listtype
+  echom l:listtype
   " we don't use lwindow to close the location list as we need also the
   " ability to resize the window. So, we are going to use lopen and lclose
   " for a better user experience. If the number of errors in a current
   " location list increases/decreases, cwindow will not resize when a new
   " updated height is passed. lopen in the other hand resizes the screen.
   if !a:0 || a:1 == 0
+    echom "first if true"
     if l:listtype == "locationlist"
+      echom "closing location list"
       lclose
     else
+      echom "closing quickfix window"
       cclose
     endif
     return
```

And here is the output of successful build (e.g. the one that should close currently open quickfix window):

```
vim-go: building dispatched ...                                                                                                                               
vim-go: SUCCESS                                                                                                                                               
0                                                                                                                                                             
0                                                                                                                                                             
first if true                                                                                                                                                 
closing location list
```

For some reasons, even though l:listtype == 0, this condition ' if l:listtype == "locationlist"' is true.

**Comment by pborzenkov** [comment]
Ok, looks like I understand what's going on here. vim casts "locationlist" to an int, since this is the type of the first comparison argument, and it becomes zero. That why the condition is true.

**Comment by pborzenkov** [comment]
@fatih Ok, so I've updated the patch to use go#list#Type and also fixed the same problem (integer arguments to go#list#Window()) in jobcontrol.vim. I don't have neovim installed, so jobcontrol.vim change is untested.

File: autoload/go/jobcontrol.vim [hunk 2]
```diff
@@ -99,9 +99,10 @@ function! s:on_exit(job_id, exit_status) abort
 
   call s:callback_handlers_on_exit(s:jobs[a:job_id], a:exit_status, std_combined)
 
+  let l:listtype = go#list#Type("locationlist")
```
Comment (fatih) [inline autoload/go/jobcontrol.vim]: Can you change this to `quickfix` as well? I know it was already like this before, but let us be consistent with the others as well. It also fixes https://github.com/fatih/vim-go/issues/1019 as well :)

**Comment by fatih** [comment]
@pborzenkov this looks good. Just a minor comment, once that is in, it's ready to be merged.

**Comment by pborzenkov** [comment]
@fatih Done.

**Comment by fatih** [comment]
Thanks @pborzenkov 👍

**Pull request merged**